src/main/java/com/aws/greengrass/device/iot/IotAuthClient.java

```diff
@@ -125,5 +135,75 @@ public boolean isThingAttachedToCertificate(Thing thing, Certificate certificate
                                 certificate.getIotCertificateId(), thing.getThingName()), e);
             }
         }
+
+        @Override
+        public void clearLocalAuthCache() {
+            certificateHashToIdMap.clear();
+            certificateIdToThingMap.clear();
+        }
+
+        @SuppressWarnings("PMD.AvoidCatchingGenericException")
+        private Optional<String> fetchActiveCertificateId(String certificatePem) {
+            VerifyClientDeviceIdentityRequest request =
+                    VerifyClientDeviceIdentityRequest.builder().clientDeviceCertificate(certificatePem).build();
+            try {
+                VerifyClientDeviceIdentityResponse response = RetryUtils.runWithRetry(SERVICE_EXCEPTION_RETRY_CONFIG,
+                        () -> clientFactory.getGreengrassV2DataClient().verifyClientDeviceIdentity(request),
+                        "verify-client-device-identity", logger);
+                return Optional.of(response.clientDeviceCertificateId());
+            } catch (InterruptedException e) {
+                logger.atError().cause(e).log("Verify client device identity got interrupted");
+                // interrupt the current thread so that higher-level interrupt handlers can take care of it
+                Thread.currentThread().interrupt();
+                throw new CloudServiceInteractionException(
+                        "Failed to verify client device identity, process got interrupted", e);
+            } catch (ValidationException | ResourceNotFoundException e) {
+                logger.atWarn().cause(e).kv("certificatePem", certificatePem)
+                        .log("Certificate doesn't exist or isn't active");
+                return Optional.empty();
+            } catch (Exception e) {
+                logger.atError().cause(e).kv("certificatePem", certificatePem)
+                        .log("Failed to verify client device identity with cloud. Check that the core device's IoT "
+                                + "policy grants the greengrass:VerifyClientDeviceIdentity permission.");
+                throw new CloudServiceInteractionException("Failed to verify client device identity", e);
+            }
+        }
+
+        private Optional<String> getCertificateIdAvailableLocally(String certificatePem) {
+            try {
+                String certHash = sha256Hash(certificatePem);
+                if (certificateHashToIdMap.containsKey(certHash)) {
+                    return Optional.of(certificateHashToIdMap.get(certHash));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **175 and 176**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **141**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).

Reply:
addressed